### PR TITLE
Datatype representation cleanup

### PR DESCRIPTION
**Pull request by pavanbalaji**
This PR contains initial precursor cleanup before adding a new datatype engine.

### CODE REVIEW

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/ofi

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets,psm2/gnu,icc/force-am-disabled,force-am-enabled/auto,netmod,no-odd-even,thread/debug,default

**Comment by hajimefu** [comment]
The OFI part of change looks good to me, but let me run the tests internally once more to make sure it doesn't break existing test cases.

**Comment by pavanbalaji** [comment]
@minsii and @hajimefu   Can comments?

**Comment by hajimefu** [comment]
@pavanbalaji Sorry, still waiting for the test results to come back (our cluster is super crowded this week...). I'll approve once I confirm the results.

File: src/mpi/datatype/looputil.c [hunk 1]
```diff
@@ -214,8 +214,8 @@ static int external32_float_convert(char *dest_buf,
  * - Assumes that the segment has been allocated.
  *
  */
-int MPIR_Segment_init(const void *buf,
-                      MPI_Aint count, MPI_Datatype handle, struct MPIR_Segment *segp)
+static int segment_init(const void *buf,
+                        MPI_Aint count, MPI_Datatype handle, struct MPIR_Segment *segp)
 {
```
Comment (minsii) [inline src/mpi/datatype/looputil.c]: Can we inline `segment_init ` ? `MPIR_Segment_alloc` may be called at perf-critical path, right ?

File: src/mpi/datatype/looputil.c [hunk 2]
```diff
@@ -365,12 +365,17 @@ int MPIR_Segment_init(const void *buf,
     return 0;
 }
 
-/* Segment_alloc
+/* MPIR_Segment_alloc
  *
```
Comment (minsii) [inline src/mpi/datatype/looputil.c]: This change looks like a pure comment fix, which is unrelated to this commit.

File: src/mpi/datatype/looputil.c [hunk 3]
```diff
@@ -365,12 +365,17 @@ int MPIR_Segment_init(const void *buf,
...
  */
-struct MPIR_Segment *MPIR_Segment_alloc(void)
+struct MPIR_Segment *MPIR_Segment_alloc(const void *buf, MPI_Aint count, MPI_Datatype handle)
 {
-    return (struct MPIR_Segment *) MPL_malloc(sizeof(struct MPIR_Segment), MPL_MEM_DATATYPE);
+    struct MPIR_Segment *segp =
+        (struct MPIR_Segment *) MPL_malloc(sizeof(struct MPIR_Segment), MPL_MEM_DATATYPE);
+
```
Comment (minsii) [inline src/mpi/datatype/looputil.c]: We might need to check return of `MPL_malloc` here.

File: src/mpi/datatype/type_flatten.c [hunk 4]
```diff
@@ -42,12 +41,9 @@ int MPIR_Type_flatten(MPI_Datatype type,
     MPIR_Assert(datatype_ptr->is_committed);
     MPIR_Assert(*array_len_p >= datatype_ptr->max_contig_blocks);
 
-    segp = MPIR_Segment_alloc();
-    err = MPIR_Segment_init(0, 1, type, segp);  /* first 0 is bufptr,
+    segp = MPIR_Segment_alloc(0, 1, type);      /* first 0 is bufptr,
                                                  * 1 is count
                                                  */
-    if (err)
-        return err;
```
Comment (minsii) [inline src/mpi/datatype/type_flatten.c]: We might want to check `segp != NULL` here. I think this fix can be in a different commit.

File: src/mpi/datatype/looputil.c [hunk 5]
```diff
@@ -365,12 +365,17 @@ int MPIR_Segment_init(const void *buf,
...
+    segment_init(buf, count, handle, segp);
+
```
Comment (minsii) [inline src/mpi/datatype/looputil.c]: Can we add a note here explaining that `segment_init ` always return 0, so we do not check the return value and handle the case where `segp` is allocated but `segment_init` fails ?

File: src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c [hunk 6]
```diff
@@ -115,15 +115,15 @@ int MPID_nem_llc_isend(struct MPIDI_VC *vc, const void *buf, int count, MPI_Data
     }
     else {
         /* See MPIDI_CH3_EagerNoncontigSend (in ch3u_eager.c) */
-        struct MPIR_Segment *segment_ptr = MPIR_Segment_alloc();
+        struct MPIR_Segment *segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
         MPIR_ERR_CHKANDJUMP(!segment_ptr, mpi_errno, MPI_ERR_OTHER, "**outofmemory");
+
 #ifndef	notdef_leak_0001_hack
```
Comment (minsii) [inline src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: Please delete the white line change.

File: src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c [hunk 7]
```diff
@@ -849,7 +847,8 @@ int llc_poll(int in_blocking_poll, llc_send_f sfnc, llc_recv_f rfnc)
                         dprintf("llc_poll,ref_count=%d,pack_buf=%p\n", req->ref_count,
                                 REQ_FIELD(req, pack_buf));
                         MPL_free(REQ_FIELD(req, pack_buf));
-			MPIR_Segment_free(seg);
+
+                        MPIR_Segment_free(seg);
                     }
```
Comment (minsii) [inline src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: Should here be unchanged ?

File: src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c [hunk 8]
```diff
@@ -1038,13 +1037,13 @@ int MPID_nem_llc_issend(struct MPIDI_VC *vc, const void *buf, int count, MPI_Dat
     }
     else {
         /* See MPIDI_CH3_EagerNoncontigSend (in ch3u_eager.c) */
-        struct MPIR_Segment *segment_ptr = MPIR_Segment_alloc();
+        segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
+
```
Comment (minsii) [inline src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: `segment_ptr ` is not declared.

File: src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_poll.c [hunk 9]
```diff
@@ -447,12 +447,14 @@ static int _mxm_process_rdtype(MPIR_Request ** rreq_p, MPI_Datatype datatype,
...
+    if (rreq->dev.segment_ptr) {
+        MPIR_Segment_free(rreq->dev.segment_ptr);
     }
-    MPIR_Segment_init(buf, count, datatype, rreq->dev.segment_ptr);
+    /* FIXME: we never seem to be freeing this segment */
+    rreq->dev.segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
+    MPIR_ERR_CHKANDJUMP1((rreq->dev.segment_ptr == NULL), mpi_errno, MPI_ERR_OTHER, "**nomem",
+                         "**nomem %s", "MPIR_Segment_alloc");
+
     rreq->dev.segment_first = 0;
```
Comment (minsii) [inline src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_poll.c]: Does this change increase overhead ? I.e., every time we have to reallocate buffer for the segment.

File: src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_send.c [hunk 10]
```diff
@@ -299,16 +299,16 @@ int MPID_nem_mxm_send(MPIDI_VC_t * vc, const void *buf, MPI_Aint count, MPI_Data
...
             last = data_sz;
             if (packsize > 0) {
                 sreq->dev.tmpbuf = MPL_malloc((size_t) packsize, MPL_MEM_BUFFER);
                 MPIR_Assert(sreq->dev.tmpbuf);
-                MPIR_Segment_init(buf, count, datatype, sreq->dev.segment_ptr);
+
+                sreq->dev.segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
+                MPIR_ERR_CHKANDJUMP1((sreq->dev.segment_ptr == NULL), mpi_errno, MPI_ERR_OTHER,
+                                     "**nomem", "**nomem %s", "MPIR_Segment_alloc");
                 MPIR_Segment_pack(sreq->dev.segment_ptr, 0, &last, sreq->dev.tmpbuf);
```
Comment (minsii) [inline src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_send.c]: The question might be unrelated to this commit: where do we free the segment ?

File: src/mpid/ch3/channels/nemesis/netmod/portals4/ptl_send.c [hunk 11]
```diff
@@ -161,15 +161,15 @@ static int send_msg(ptl_hdr_data_t ssend_flag, struct MPIDI_VC *vc, const void *
...
+        sreq->dev.segment_ptr = MPIR_Segment_alloc(buf, count, datatype);
         MPIR_ERR_CHKANDJUMP1(sreq->dev.segment_ptr == NULL, mpi_errno, MPI_ERR_OTHER, "**nomem", "**nomem %s", "MPIR_Segment_alloc");
-        MPIR_Segment_init(buf, count, datatype, sreq->dev.segment_ptr);
         sreq->dev.segment_first = 0;
         sreq->dev.segment_size = data_sz;
 
         last = sreq->dev.segment_size;
         sreq->dev.iov_count = MPL_IOV_LIMIT;
         MPIR_Segment_to_iov(sreq->dev.segment_ptr, sreq->dev.segment_first, &last, sreq->dev.iov, &sreq->dev.iov_count);
+        MPIR_Segment_free(sreq->dev.segment_ptr);
```
Comment (minsii) [inline src/mpid/ch3/channels/nemesis/netmod/portals4/ptl_send.c]: Looks like you fixed the unreleased segment issue in portals. Could you please also fix the same issue in mxm ? Besides, this fix should be put into a separate commit.

File: src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c (hunk 7 above)
Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: We are allocating the segment, so it needs to be freed, right?  What's the issue?

Comment (minsii) [reply src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: This change is a simple format clean up, right ?

File: src/mpi/datatype/looputil.c (hunk 1 above)
Comment (pavanbalaji) [reply src/mpi/datatype/looputil.c]: Done.

File: src/mpi/datatype/looputil.c (hunk 2 above)
Comment (pavanbalaji) [reply src/mpi/datatype/looputil.c]: Done.

File: src/mpi/datatype/looputil.c (hunk 3 above)
Comment (pavanbalaji) [reply src/mpi/datatype/looputil.c]: Done.

File: src/mpi/datatype/type_flatten.c (hunk 4 above)
Comment (pavanbalaji) [reply src/mpi/datatype/type_flatten.c]: Done.  Btw, this file will be completely rewritten in a follow-up commit that I have coming.

File: src/mpi/datatype/looputil.c (hunk 5 above)
Comment (pavanbalaji) [reply src/mpi/datatype/looputil.c]: Instead, I've changed the return type of `segment_init` to `void`.

File: src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c (hunk 7 above)
Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: Oh, yes.  Sorry, I was looking at the wrong commit.

File: src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c (hunk 8 above)
Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: Fixed.

File: src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_poll.c (hunk 9 above)
Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_poll.c]: It does, slightly.  But the bigger cost is in the segment initialization, which we are not avoiding. 
 Also, this is the derived datatype code, which is very slow in the current model.  Once the new derived datatype code is ready, we will not have these alloc/frees in that, which will be much faster.

File: src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_send.c (hunk 10 above)
Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/netmod/mxm/mxm_send.c]: Fixed.

File: src/mpid/ch3/channels/nemesis/netmod/portals4/ptl_send.c (hunk 11 above)
Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/netmod/portals4/ptl_send.c]: Done.

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/mxm

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 12]
```diff
@@ -707,20 +707,26 @@ MPL_STATIC_INLINE_PREFIX
...
-        MPIR_Segment_manipulate(&dt_seg, 0, &dt_size,
+        dt_seg = MPIR_Segment_alloc();
+        MPIR_Segment_init(NULL, 1, dt_datatype, dt_seg);
+        MPIR_Segment_manipulate(dt_seg, 0, &dt_size,
                                 MPIDI_OFI_contig_count_block,
                                 NULL, NULL, NULL, NULL, (void *) &params);
+        MPIR_Segment_free(dt_seg);
         count1 = params.count;
+
         params.count = 0;
```
Comment (minsii) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: White line change

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 13]
```diff
@@ -707,20 +707,26 @@ MPL_STATIC_INLINE_PREFIX
...
-        MPIR_Segment_manipulate(&dt_seg, 0, &dt_size,
+        dt_seg = MPIR_Segment_alloc();
+        MPIR_Segment_init(NULL, dtc, dt_datatype, dt_seg);
+        MPIR_Segment_manipulate(dt_seg, 0, &dt_size,
                                 MPIDI_OFI_contig_count_block,
                                 NULL, NULL, NULL, NULL, (void *) &params);
+        MPIR_Segment_free(dt_seg);
         count2 = params.count;
+
         if (count2 == 1) {      /* Contiguous */
```
Comment (minsii) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h]: While line change

File: src/mpi/datatype/dataloop/dataloop_debug.c [hunk 14]
```diff
@@ -0,0 +1,219 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
```
[the remark this answers is not in the document]
Comment (pavanbalaji) [reply src/mpi/datatype/dataloop/dataloop_debug.c]: It's not.  It's just moved from a different location.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 12 above)
Comment (pavanbalaji) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: Fixed locally.  Will push once the tests complete.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 13 above)
Comment (pavanbalaji) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: Fixed locally.  Will push once the tests complete.

**Comment by pavanbalaji** [comment]
@minsii The PR is ready for your review again.

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 12 above)
Comment (minsii) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: Is this not pushed ?

Comment (pavanbalaji) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h]: It is.  The above code shows as "outdated".

File: src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c (hunk 7 above)
Comment (minsii) [reply src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: This line is still shown as changed. Is it a tab -> space format change ?

Comment (pavanbalaji) [reply src/mpid/ch3/channels/nemesis/netmod/llc/llc_send.c]: Fixed.

**Comment by pavanbalaji** [comment]
@minsii I think have addressed all of your comments.  Can you take a look at this PR again?

**Comment by pavanbalaji** [comment]
Added signoffs and rebased on to the latest `mpich/master`.  Rerunning tests once more before merging.

test:jenkins/ch3/most
test:jenkins/ch4/most
